Store run info in tags

acroz opened the pull request:
It's planned to have less information about runs be stored as top-level attributes on runs and push it into tags. This will make the models simpler and remove some inconsistencies with how data is currently structured.

This PR adds additional tags to the system-reserved `mlflow.` namespace and updates `mlflow.tracking.fluent` to write them on run creation. Specifically, the following tags have been added:

- `mlflow.source.type` - currently stored as the `source_type` attribute of the run
- `mlflow.source.name` - currently stored as the `source_name` attribute of the run
- `mlflow.source.git.commit` - currently stored as the `source_version` attribute of the run
- `mlflow.project.entryPoint` - currently stored as the `entry_point_name` attribute of the run

In addition, the following tags have been renamed to match the above pattern:

- `mlflow.source.git.branch` - previously `mlflow.gitBranchName`
- `mlflow.source.git.repoURL` - previously `mlflow.gitRepoURL`

The existing top-level attributes of the runs continue to be set, as do the older versions of the renamed tags, for preservation of existing functionality. It is anticipated that a future PR (before the v1.0 release) will remove these top-level attributes in favour of reading from tags, at which time users will hopefully have all or most of their data in tags for use by the UI and other tools.

This PR is dependent on PR #913, which implements tests of the existing `mlflow.tracking.fluent.start_run` functionality. See https://github.com/facultyai/mlflow/compare/fluent-start-run-tests...facultyai:run-info-in-tags for a preview of the changes introduced by this PR in isolation.

acroz committed:
Add tests for fluent.start_run for existing run

acroz committed:
Test creation of new runs in start_run

acroz committed:
Test raising exception on deleted run retrieval

acroz committed:
Test logic for registering parent run

acroz committed:
Update tests to original implementation

acroz committed:
Refactor mlflow.tracking.fluent.start_run

acroz committed:
Get tag names from utils

acroz committed:
Change tags for git branch and repo URL but keep adding old ones

acroz committed:
Modify tag validation logic

This change makes sure that the desired tags actually exist and will
also not fail if additional tags are added, which is anticipated.

acroz committed:
Store git commit as a tag

acroz committed:
Store source name and type in tags

acroz committed:
Store project entry point in tags

acroz commented:
@aarondav Can we rerun the tests on this - AFAICT it looks like they've failed for an unrelated reason.

acroz committed:
Merge branch 'master' of github.com:mlflow/mlflow into fluent-start-run-tests

aarondav commented on mlflow/utils/mlflow_tags.py:
```diff
@@ -1,18 +1,27 @@
 MLFLOW_DOCKER_IMAGE_NAME = "mlflow.docker.image.name"
 MLFLOW_DOCKER_IMAGE_ID = "mlflow.docker.image.id"
+
+LEGACY_MLFLOW_GIT_BRANCH_NAME = "mlflow.gitBranchName"
```
Could we add a comment on these (e.g., these were the tag names prior to 0.9.0, and we want to keep backwards-compatibility for at least one release)?

acroz replied:
Done.

aarondav commented on mlflow/utils/mlflow_tags.py:
```diff
@@ -1,18 +1,27 @@
 MLFLOW_RUN_NAME = "mlflow.runName"
-MLFLOW_GIT_BRANCH_NAME = "mlflow.gitBranchName"
-MLFLOW_GIT_REPO_URL = "mlflow.gitRepoURL"
+MLFLOW_SOURCE_TYPE = "mlflow.source.type"
```
Currently, source type is typed -- NOTEBOOK / JOB / PROJECT / LOCAL / UNKNOWN. With this change, we will lose the documentation on what "built-in" source types are, and what expected values look like. It seems like we should probably have a docs page where we define the default tags for MLflow, similar to our REST API docs.

acroz replied:
As discussed, I've added this to the REST API docs for now. I'm not sure of meaning of some of the Databricks-specific tags, though - can you have a look and provide descriptions?

aarondav commented on tests/store/test_sqlalchemy_store.py:
```diff
@@ -307,7 +308,7 @@ def test_create_run(self):
         self.assertEqual(len(actual.data.tags), 3)
 
-        name_tag = models.SqlTag(key='mlflow.runName', value='booyya').to_mlflow_entity()
+        name_tag = models.SqlTag(key=MLFLOW_RUN_NAME, value='booyya').to_mlflow_entity()
```
This change is intended to be staged across two releases, but it may make sense to make more of the changes that 'assume source_type is in tags only'. These would include:

- Removing these parameters from the MlflowClient tracking API (e.g., mlflow/projects/__init__.py#create_run would have to embed these in tags instead).
- Mentioning in the protos that these fields are deprecated and moved to tags.
- Changing the JavaScript to read first from the tag and then from the field.

acroz replied:
We agreed that a subsequent PR should implement these.

acroz committed:
Merge branch 'master' of github.com:mlflow/mlflow into run-info-in-tags

acroz committed:
Reorder tags and add comment about legacy tag deprecation

acroz committed:
Rename tag constant name

acroz committed:
Remove values for project env tag from `mlflow.utils.mlflow_tags`

Currently, all other values in `mlflow.utils.mlflow_tags` are tag keys,
and there is nothing to indicate in that module that `MLFLOW_DOCKER` and
`MLFLOW_CONDA` are actually intended for use as tag values.

At some point, an enumeration of valid values for this tag could be
created, but since that seems outside the scope of this PR and these tag
values are only used in a single location, I've just baked them in place
for the time being.

acroz committed:
WIP: Document meaning of tags.

Waiting for @aarondav to confirm the meaning of some of the
Databricks-specific tags before completing the second table.

acroz committed:
Add deprecation notices to protobuf file

acroz committed:
Regenerate protobuf files

acroz committed:
Tweak deprecation message

acroz committed:
Point users to the REST API docs for information on tag meaning

aarondav commented on docs/source/rest-api.rst:
```diff
@@ -1029,6 +1029,53 @@ Tag for a run.
+| ``mlflow.docker.image.id``    | ID of the Docker image used to execute this run.                                       |
++-------------------------------+----------------------------------------------------------------------------------------+
+
+The following tags are set when running MLflow in Databricks:
```
We can probably just remove the documentatino for these -- when we move the databricks stuff to an external plugin, they should be documented there.

aarondav commented on docs/source/rest-api.rst:
```diff
@@ -1029,6 +1029,53 @@ Tag for a run.
 | value      | ``STRING`` | The tag value. |
 +------------+------------+----------------+
 
+Tag keys that start with ``mlflow.`` are reserved for internal use. The following tags are set
```
cc @stbof Here we are documenting MLflow system tags. We are currently planning on putting them in the REST API documentation as part of `RunTag`. Any thoughts on this location, or places where we should link to these? (Also any comments on the docs themselves?)

stbof replied:
@aarondav I think there should be a link from SetTag to the list of disallowed tag keys.

aarondav commented:
LGTM - let's just remove the databricks-specific tags for now, and we can merge. If there are other docs follow-ups, can probably do those in a follow-up.

aarondav commented on docs/source/rest-api.rst:
```diff
@@ -1029,6 +1029,53 @@ Tag for a run.
 | value      | ``STRING`` | The tag value. |
 +------------+------------+----------------+
 
+Tag keys that start with ``mlflow.`` are reserved for internal use. The following tags are set
```
cc @stbof Here we are documenting MLflow system tags. We are currently planning on putting them in the REST API documentation as part of `RunTag`. Any thoughts on this location, or places where we should link to these? (Also any comments on the docs themselves?)

aarondav committed:
Remove databricks-specific tags

aarondav commented:
Btw, went ahead and removed the databricks-specific tags to kick off a build.
